Updates to guide

sdaschner opened the pull request:
Creating a PR based on the conversations the last days. It was easier to create a PR rather than issues; I don't expect this to be merged immediately, rather, we can comment on these and I can update / refine.


In general, I'd advocate for:

- for using a thin WAR approach where Maven builds the application only, that is then deployed to Liberty, either in a Docker container, or in the guide via the `liberty` plugin.
- a minimalistic approach in the `pom.xml`

I've made following changes:

- Simplified the `pom.xml`
- Updated the Maven plugin to `3.x`, and thus removed all configuration that is now a default (e.g. no need for server config location, or the Maven parent dependency)
- Change the examples to use the `dev` goal of the plugin for reloads; this might cause that we only publish after the `3.0` version has been published.
- Changed Docker example to only contain plain base image and adding the built WAR file
- Tests changed to IT, automatically ignored from surefire by Maven convention
- Removed metrics over HTTP resource, possible by server config element `<mpMetrics/>`
- Updated MicroProfile Metrics to 2.0
- Moved the structure around, e.g. moved Docker before the (optional) step of producing a standalone JAR

Still open:

- The standalone JAR approach didn't work OOTB with my changes to the plugin version `3.0`; we need to revise this, or leave the section out for now (in the getting started guide)

scottkurz commented on start/pom.xml:
```diff
@@ -161,126 +84,52 @@
-                            <systemPropertyVariables>
-                                <liberty.test.port>${http.port}</liberty.test.port>
-                                <war.name>${app.name}</war.name>
-                            </systemPropertyVariables>
```
After removing this how were you thinking the port gets into the client ITs (doing `String port = System.getProperty("liberty.test.port");` )

sdaschner replied:
`System.getProperty("liberty.test.port", "9080");` for the default, which always works in my cases and, if necessary, override it on executing the Maven ITs with `-Dliberty.test.port=9081` ...

scottkurz replied:
What I really meant to say is that the previous approach parameterized the client ITs while parameterizing the server at the same time, via this config:

                    <bootstrapProperties>
                        <default.http.port>${http.port}</default.http.port>
                        <default.https.port>${https.port}</default.https.port>
                        <app.context.root>${app.name}</app.context.root>
                    </bootstrapProperties>

If you passed a 9081 value for `liberty.test.port` how would the server pick up the configuration to launch ?

sdaschner replied:
I see your point. This config was actually something I'd remove from the getting-started-guide at least, since it overloads the user with something that is IMO too detailed for that abstraction level already.

So, I'd keep the basic sample lean, and maybe only point to the advanced configuration, maybe to a different guide or docs. How we could make it work in a more leaner way without touching the pom.xml is to define the server.xml with `<httpEndpoint httpPort="${default.http.port}" ... />`, and override the var via the bootstrap properties when running Maven: `mvn ... -DbootstrapProperties=default.http.port=9081`.

But, I'd actually keep it out of the first guide to not confuse. For getting started users it's IMO better to start with a lean config, since it's not required to run Liberty. Otherwise, developers might assume that all that config is required to get started, which is not true for our lean technology.

gkwan-ibm commented:
I will review the changes. By the way, we are modifying this guide too. Some conflicts are expected.

gkwan-ibm commented:
**README.adoc**
- [ ] From the line 96 to 104, the paragraph "As part of this phase, ...loose config." was moved after the statment "The `liberty:run-server` command...". It is not right because the original meaning was changed. i.e. It is not the `liberty:run-server` command to download and install the Liberty runtime but the `package` command does.
- [ ] The following output line at line 197 should be updated to `mpHealth-2.0` instead of `mpHealth-1.0`
   - `[INFO] [AUDIT] CWWKF0012I: The server installed the following features: [mpHealth-1.0].`
- [ ] The section "Starting and stopping the Open Liberty server in the background" was moved ahead. I think it is not enough. It should be moved before the section "Updating the server configuration without restarting the server", because we introduce `liberty:dev` starting the "Updating the server configuration without restarting the server" section.
- [ ] At the end of the "Checking the Open Liberty server logs" section, need to ask readers to stop the dev mode by `liberty:stop-server` or CTRL+C, so that they can proceed to next section.

I cannot process to review because the `mvn package` fails (the travis continuous-integration test failed with the same following error) 
```
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  2.947 s
[INFO] Finished at: 2019-08-08T17:18:26-04:00
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal net.wasdev.wlp.maven.plugins:liberty-maven-plugin:3.0.M1:dev (default-cli) on project getting-started: Unable to execute mojo: Compilation failure: 
[ERROR] /Users/gkwan/tasks/CNAI/guides/gkwan-ibm/guide-getting-started/finish/src/main/java/io/openliberty/sample/system/SystemReadinessCheck.java:[20,46] package org.eclipse.microprofile.config.inject does not exist
[ERROR] /Users/gkwan/tasks/CNAI/guides/gkwan-ibm/guide-getting-started/finish/src/main/java/io/openliberty/sample/system/SystemResource.java:[33,35] cannot find symbol
[ERROR]   symbol:   method monotonic()
[ERROR]   location: @interface org.eclipse.microprofile.metrics.annotation.Counted
[ERROR] /Users/gkwan/tasks/CNAI/guides/gkwan-ibm/guide-getting-started/finish/src/main/java/io/openliberty/sample/system/SystemReadinessCheck.java:[31,6] cannot find symbol
[ERROR]   symbol:   class ConfigProperty
[ERROR]   location: class io.openliberty.sample.system.SystemReadinessCheck
[ERROR] -> [Help 1]
[ERROR] 
```

sdaschner commented:
@gkwan-ibm Thanks for your updates, finally getting back:

> From the line 96 to 104, the paragraph "As part of this phase, ...loose config." was moved after the statment "The liberty:run-server command...". It is not right because the original meaning was changed. i.e. It is not the liberty:run-server command to download and install the Liberty runtime but the package command does.
Actually not in the way the pom.xml is structured now. I don't define a parent anymore and if you now only run `mvn package`, only the WAR is built, which is a huge benefit, since that build runs much faster.

> The following output line at line 197 should be updated to mpHealth-2.0 instead of mpHealth-1.0
[INFO] [AUDIT] CWWKF0012I: The server installed the following features: [mpHealth-1.0].
Good catch! Updated.

> The section "Starting and stopping the Open Liberty server in the background" was moved ahead. I think it is not enough. It should be moved before the section "Updating the server configuration without restarting the server", because we introduce liberty:dev starting the "Updating the server configuration without restarting the server" section.
You're right, this structure makes more sense. Moved.

> I cannot process to review because the mvn package fails (the travis continuous-integration test failed with the same following error)
Oops, you're right. Fixed.


Also, I've changed the occurrences of the `-server` goals to their shorter counterparts, e.g. `run` instead of `run-server`.

scottkurz commented:
BTW @sdaschner - not sure if you saw it but I sent a PR (to your fork) to fix the compile errors.

gkwan-ibm commented:
The dependencies should be:
```
        <dependency>
            <groupId>javax.ws.rs</groupId>
            <artifactId>javax.ws.rs-api</artifactId>
            <version>2.1.1</version>
            <scope>provided</scope>
        </dependency>
        <dependency>
            <groupId>javax.json</groupId>
            <artifactId>javax.json-api</artifactId>
            <version>1.1.4</version>
            <scope>provided</scope>
        </dependency>
        <dependency>
            <groupId>javax.enterprise</groupId>
            <artifactId>cdi-api</artifactId>
            <version>2.0</version>
            <scope>provided</scope>
        </dependency>
        <dependency>
            <groupId>javax.enterprise</groupId>
            <artifactId>cdi-api</artifactId>
            <version>2.0</version>
            <scope>provided</scope>
        </dependency>
        <dependency>
            <groupId>org.eclipse.microprofile.metrics</groupId>
            <artifactId>microprofile-metrics-api</artifactId>
            <version>1.1</version>
            <scope>provided</scope>
        </dependency>
        <dependency>
            <groupId>org.eclipse.microprofile.health</groupId>
            <artifactId>microprofile-health-api</artifactId>
            <version>2.0</version>
            <scope>provided</scope>
        </dependency>
        <dependency>
            <groupId>org.eclipse.microprofile.config</groupId>
            <artifactId>microprofile-config-api</artifactId>
            <version>1.3</version>
            <scope>provided</scope>
        </dependency>```
Note: The above is not all.
- metrics api need 1.1 (will need to update the app and to use 2.0)
- 2 more dependencies cdi-api and microprofile-config-api are needed

gkwan-ibm commented:
>     * metrics api need 1.1 (will need to update the app and to use 2.0)
> 
>     * 2 more dependencies cdi-api and microprofile-config-api are needed

Both were addressed in the PR
